DEVELOPMENT.md

```diff
@@ -0,0 +1,307 @@
+  done
+fi
+```
+
+## Testing strategy
+
+### Running tests and linting
+
+**Plugin Tester** - Run all tests:
+
+```bash
+docker run -it --rm -v "$PWD:/plugin:ro" buildkite/plugin-tester
+```
+
+**Plugin Linter** - Validate plugin structure:
+
+```bash
+# Replace 'your-plugin-name' with your actual plugin name
+docker run -it --rm -v "$PWD:/plugin:ro" buildkite/plugin-linter --id your-plugin-name --path /plugin
+```
+
+**ShellCheck** - Static analysis for shell scripts:
+
+```bash
+shellcheck hooks/* tests/* lib/*.bash lib/modules/* lib/providers/*
+```
+
+### Unit testing with BATS
+
+Test individual functions:
+
+```bash
+@test "validates required config" {
+  export BUILDKITE_PLUGIN_MYPLUGIN_API_TOKEN=""
+  run validate_required_config "API token" "${BUILDKITE_PLUGIN_MYPLUGIN_API_TOKEN}"
+  assert_failure 1  # ensure it fails with exit code 1
+  assert_output --partial "API token is required"
+}
+```
+
+### Integration testing
+
+Test full plugin execution with realistic scenarios:
+
+```bash
+@test "handles missing dependencies gracefully" {
+  # Mock missing command
+  run hooks/command
+  assert_failure
+  assert_output --partial "Missing required dependencies"
+}
+```
+
+### Test structure
+
+```
+tests/
+└── command.bats           # Plugin functionality tests
+```
+
+## Performance considerations
+
+### Avoid expensive operations in early hooks
+
+```bash
+# Bad - slow network call in environment hook
+validate_api_connectivity "${api_url}"
+
+# Good - defer expensive operations to command hook
+export PLUGIN_API_URL="${api_url}"
+```
+
+### Cache expensive lookups
+
+Instead of doing:
+```bash
+get_account_id() {
+  aws sts get-caller-identity --query Account --output text
+}
+
+# somewhere else
+ACCOUNT_ID="$(get_account_id)"
+```
+
+Cache the result of the call:
+```bash
+get_account_id() {
+  # Cache account ID lookup
+  if [[ -z "${CACHED_ACCOUNT_ID:-}" ]]; then
+    CACHED_ACCOUNT_ID=$(aws sts get-caller-identity --query Account --output text)
+    export CACHED_ACCOUNT_ID
+  fi
+
+  echo "$CACHED_ACCOUNT_ID"
+}
+
+# somewhere else
+ACCOUNT_ID="$(get_account_id)"
+```
+
+## Security considerations
+
+### Validate input
+
+```bash
+# Prevent injection attacks
+if [[ ! "${region}" =~ ^[a-z0-9-]+$ ]]; then
+  log_error "Invalid region format: ${region}"
+  exit 1
+fi
+```
+
+### Use secure secret handling
+
+Do not reference secrets directly in options or in the pipeline or with special names.
+
+The following pipeline will get `SECRET_API_TOKEN` interpolated in the step that does the pipeline upload and not correctly redacted afterwards (without modifying the agent's configuration):
+
+```yaml
+plugins:
+  - myplugin#v1.0.0:
+      my-secret-variable: $SECRET_API_TOKEN
+```
+
+Instead, take an environment variable **name** and take advantage of bash's dereference functionality:
+
+```yaml
+plugins:
+  - myplugin#v1.0.0:
+      secret-variable-name: MY_VARIABLE
+```
+
+And in the code:
+```bash
+SECRET_TOKEN="${!MY_VARIABLE}"
+```
```

Comment:
Actually, this will not work as `${!VAR}` will result in the value of the variable that has the name contained in `VAR`, a better (and correct) example would be:

````suggestion
And in the code:
```bash
SECRET_VARIABLE_NAME="$(plugin_read_config secret-variable-name)"
SECRET_TOKEN="${!SECRET_VARIABLE_NAME}"
```
````